fuel-crypto/src/secp256/backend/k1/k256.rs

```diff
@@ -0,0 +1,147 @@
+use crate::{
+    message::Message,
+    secp256::{
+        signature_format::{
+            decode_signature,
+            encode_signature,
+            RecoveryId as SecpRecoveryId,
+        },
+        PublicKey,
+    },
+    Error,
+};
+
+#[cfg(feature = "random")]
+use crate::SecretKey;
+
+use k256::{
+    ecdsa::{
+        RecoveryId,
+        VerifyingKey,
+    },
+    EncodedPoint,
+};
+
+#[cfg(feature = "random")]
+use rand::{
+    CryptoRng,
+    RngCore,
+};
+
+/// Generates a random secret key
+#[cfg(feature = "random")]
+pub fn random_secret(rng: &mut (impl CryptoRng + RngCore)) -> SecretKey {
+    k256::SecretKey::random(rng).into()
+}
+
+/// Derives the public key from a given secret key
+pub fn public_key<SK: Into<k256::SecretKey>>(secret: SK) -> PublicKey {
+    let sk: k256::SecretKey = secret.into();
+    let sk: ecdsa::SigningKey<k256::Secp256k1> = sk.into();
+    let vk = sk.verifying_key();
+    vk.into()
+}
+
+/// Sign a given message and compress the `v` to the signature
+///
+/// The compression scheme is described in
+/// <https://github.com/FuelLabs/fuel-specs/blob/master/src/protocol/cryptographic-primitives.md>
+pub fn sign<SK: Into<k256::SecretKey>>(secret: SK, message: &Message) -> [u8; 64] {
```

Comment:
maybe we should use an actual trait interface in front of these backends so that it's less likely users of this crate get unexpected errors when switching between std and nostd?

Reply:
I considered it, but I really don't see what extra value using a trait here would bring. Any errors would already be caught by the type system, or at least the tests. I wont mind wrapping them in a trait here if you feel like it's better like that.

Reply:
Also, the types used by `public_key` and `sign` functions are not the same for both modules, as they use `Into<...>` with a type from the corresponding backing crate, although that could be changed as well.

Reply:
It would probably be better if we had our own types on the public interfaces, and then implemented conversion to various backends as needed based on what's compiled in. Otherwise if one of these crypto libs changes their types or interfaces it could cause a lot of breaking changes in other places.

Reply:
Agreed, although I find it unlikely to actually cause breakage, but it's true that upgrading an underlying crypto library is a breaking change in the current implementation. I'll fix that.

Reply:
Ok, I've fixed this now for `secp256k1`. For `secp256r1` we currently have only a single implementation, which still exposes some types from the underlying crypto lib under `#[cfg(feature = "test-helpers")]`, which I think is fine.

Reply:
Which commit has this fix? I'm not seeing it

Reply:
f52fe821ce585b43a43cd2e8c03734f083ef0985 and 053fa238f122214ad79670de54029116434bd8c7. Forgot to push those, sorry.